File: leetcode/106.从中序与后序遍历序列构造二叉树.cpp
/*
 * @lc app=leetcode.cn id=106 lang=cpp
 *
 * [106] 从中序与后序遍历序列构造二叉树
 */
#include <vector>
#include <iostream>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* dfs(int low1, int high1, int low2, int high2,const vector<int>& inorder, const vector<int>& postorder) {
        if (low1 > high1) {
            return NULL;
        }
        if (low1 == high1) {
            return new TreeNode(inorder[low1]);
        }
        int rootVal = postorder[high2];
        int rootIdx = low1;
        for (; rootIdx <= high1;) {
            if (inorder[rootIdx] != rootVal) {
                ++rootIdx;
            } else {
                break;
            }
        }
        TreeNode* root = new TreeNode(rootVal);
        int rightLen = high1 - rootIdx;
        root->left = dfs(low1, rootIdx - 1, low2, high2 - rightLen - 1, inorder, postorder);
        root->right = dfs(rootIdx + 1, high1,  high2 - rightLen - 1, high2 - 1, inorder, postorder);
        return root;
    }

    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return dfs(0, inorder.size() - 1, 0, inorder.size() - 1, inorder, postorder);
    }
};
// @lc code=end

int main() {
    vector<int> inorder = {1,2,3,4};
    vector<int> postorder = {3,2,4,1};
    Solution{}.buildTree(inorder, postorder);
    cout << "je" << endl;
}

File: python/106.从中序与后序遍历序列构造二叉树.py
#
# @lc app=leetcode.cn id=106 lang=python3
#
# [106] 从中序与后序遍历序列构造二叉树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if len(inorder) == 0 or len(postorder) == 0:
            return None
        delimiter = 0
        val = postorder[len(postorder) - 1]

        for i in range(len(inorder)):
            if inorder[i] == val:
                delimiter = i
                break
            
        root = TreeNode(val)

        inLeft = inorder[:delimiter]
        inRight = inorder[delimiter + 1:]

        postLeft = postorder[:len(inLeft)]
        postRight = postorder[len(inLeft): -1]

        root.left = self.buildTree(inLeft, postLeft)
        root.right = self.buildTree(inRight, postRight)
        return root
# @lc code=end

